test/libyul/yulOptimizerTests/fullSuite/abi2.yul

```diff
@@ -0,0 +1,1218 @@
+			}
+		
+			function cleanup_assert_t_uint256(value) -> cleaned {
+				cleaned := value
+			}
+		
+			function cleanup_revert_t_address(value) -> cleaned {
+				cleaned := cleanup_assert_t_uint160(value)
+			}
+		
+			function cleanup_revert_t_address_payable(value) -> cleaned {
+				cleaned := cleanup_assert_t_uint160(value)
+			}
+		
+			function cleanup_revert_t_bool(value) -> cleaned {
+				cleaned := iszero(iszero(value))
+			}
+		
+			function cleanup_revert_t_bytes32(value) -> cleaned {
+				cleaned := value
+			}
+		
+			function cleanup_revert_t_contract$_Module_$1038(value) -> cleaned {
+				cleaned := cleanup_assert_t_address(value)
+			}
+		
+			function cleanup_revert_t_enum$_Operation_$1949(value) -> cleaned {
+				if iszero(lt(value, 3)) { revert(0, 0) } cleaned := value
+			}
+		
+			function cleanup_revert_t_uint256(value) -> cleaned {
+				cleaned := value
+			}
+		
+			function convert_t_contract$_GnosisSafe_$710_to_t_address_payable(value) -> converted {
+				converted := convert_t_contract$_GnosisSafe_$710_to_t_uint160(value)
+			}
+		
+			function convert_t_contract$_GnosisSafe_$710_to_t_uint160(value) -> converted {
+				converted := cleanup_assert_t_uint160(value)
+			}
+		
+			function convert_t_contract$_Module_$1038_to_t_address(value) -> converted {
+				converted := convert_t_contract$_Module_$1038_to_t_uint160(value)
+			}
+		
+			function convert_t_contract$_Module_$1038_to_t_uint160(value) -> converted {
+				converted := cleanup_assert_t_uint160(value)
+			}
+		
+			function convert_t_enum$_Operation_$1949_to_t_uint8(value) -> converted {
+				converted := cleanup_assert_t_enum$_Operation_$1949(value)
+			}
+		
+				function copy_calldata_to_memory(src, dst, length) {
+					calldatacopy(dst, src, length)
+					// clear end
+					mstore(add(dst, length), 0)
+				}
+			
+				function copy_memory_to_memory(src, dst, length) {
+					let i := 0
+					for { } lt(i, length) { i := add(i, 32) }
+					{
+						mstore(add(dst, i), mload(add(src, i)))
+					}
+					if gt(i, length)
+					{
+						// clear end
+						mstore(add(dst, length), 0)
+					}
+				}
+			
+			function round_up_to_mul_of_32(value) -> result {
+				result := and(add(value, 31), not(31))
+			}
+			}
+// ----
+// fullSuite
+// {
+//     {
+//         let _1 := 1
+//         let _2 := mload(_1)
+//         let _3 := 0
+//         let _1017 := mload(_3)
+//         let abi_decode_value0_156 := _3
+//         if slt(sub(_2, _1017), 64)
+//         {
+//             revert(_3, _3)
+//         }
+//         {
+//             abi_decode_value0_156 := and(calldataload(_1017), 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF)
+//         }
+//         sstore(_3, abi_decode_value0_156)
+//         let _1019 := mload(8)
+//         let _1021 := mload(7)
+//         let abi_decode_value0_57 := _3
+//         let abi_decode_value1_58 := _3
+//         let abi_decode_value2_59 := _3
+//         let abi_decode_value3 := _3
+//         let abi_decode_value4 := _3
```

Comment:
This prompts another optimizer stage:

Move all variable declarations as far down as possible.

Reply:
Although it might be that this could solve itself if we apply the flattener.

Reply:
Yes, this cries for the flattener... Not sure why one wouldn't always try to flatten everything as much as possible actually... Just because of potential stack height issues?

Reply:
Yep, it worked: https://github.com/ethereum/solidity/pull/5761/files?utf8=%E2%9C%93&diff=unified&w=1#diff-67de40953aaa194c53eda8352f19fdd3R1087